sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MarathonConstraintParser.java

```diff
@@ -0,0 +1,321 @@
+     *     or a {@code null} or empty value if no constraint is defined
+     * @throws IOException if {@code marathonConstraints} couldn't be parsed, or if the parsed
+     *     content isn't valid or supported
+     */
+    public static PlacementRule parse(String marathonConstraints) throws IOException {
+        if (marathonConstraints == null || marathonConstraints.isEmpty()) {
+            // nothing to enforce
+            return new PassthroughRule();
+        }
+        List<List<String>> rows = splitConstraints(marathonConstraints);
+        if (rows.size() == 1) {
+            return parseRow(rows.get(0));
+        } else {
+            List<PlacementRule> rowRules = new ArrayList<>();
+            for (List<String> row : rows) {
+                rowRules.add(parseRow(row));
+            }
+            return new AndRule(rowRules);
+        }
+    }
+
+    /**
+     * Converts the provided marathon constraint entry to a PlacementRule.
+     *
+     * @param row a list with size 2 or 3
+     * @throws IOException if the provided constraint entry is invalid
+     */
+    private static PlacementRule parseRow(List<String> row) throws IOException {
+        if (row.size() < 2 || row.size() > 3) {
+            throw new IOException(String.format(
+                    "Invalid number of entries in rule. Expected 2 or 3, got %s: %s",
+                    row.size(), row));
+        }
+        // row fields: fieldname, operatorname[, parameter]
+        final String fieldName = row.get(0);
+        final String operatorName = row.get(1);
+        final Optional<String> parameter = Optional.ofNullable(row.size() >= 3 ? row.get(2) : null);
+        Operator operator = SUPPORTED_OPERATORS.get(operatorName);
+        if (operator == null) {
+            throw new IOException(String.format(
+                    "Unsupported operator: '%s' in constraint: %s " +
+                    "(expected one of: UNIQUE, CLUSTER, GROUP_BY, LIKE, UNLIKE, or MAX_PER)",
+                    operatorName, row));
+        }
+        PlacementRule rule = operator.run(fieldName, operatorName, parameter);
+        LOGGER.info("Marathon-style row '{}' resulted in placement rule: '{}'", row, rule);
+        return rule;
+    }
+
+    /**
+     * Splits the provided marathon constraint statement into elements. Doesn't do any validation
+     * on the element contents.
+     * @param marathonConstraints
+     * @return
+     * @throws IOException
+     */
+    @VisibleForTesting
+    static List<List<String>> splitConstraints(String marathonConstraints) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        // The marathon doc uses a format like: '[["a", "b", "c"], ["d", "e"]]'
+        // Meanwhile the marathon web interface uses a format like: 'a:b:c,d:e'
+        try {
+            // First try: ["a", "b", "c"] (not technically correct but lets be lenient here)
```

Comment:
This comment is a little difficult to understand, I'd either flesh it out more or remove it.